[XLA] Remove unused dlfcn.h and implement sincos[f] for MSVC

rongjiecomputer opened the pull request:
Split from #15310.

#15213

rongjiecomputer committed:
[XLA] Remove unused dlfcn.h and implement sincos[f] for MSVC

tensorflow-jenkins commented:
Can one of the admins verify this patch?

tatatodd commented on tensorflow/compiler/xla/service/cpu/simple_orc_jit.cc:
```diff
@@ -41,6 +40,19 @@ limitations under the License.
 #include "tensorflow/compiler/xla/types.h"
 #include "tensorflow/core/platform/logging.h"
 
+#ifdef _MSC_VER
```
I'd prefer not cluttering up this file, and instead put these platform-specific (missing) functions in a separate file.  Let's see what @sanjoy thinks about that.

rongjiecomputer replied:
> I'd prefer not cluttering up this file, and instead put these platform-specific (missing) functions in a separate file.

Agree.

sanjoy replied:
+1 to this.

rongjiecomputer replied:
@sanjoy Any idea where should I put them though? I am still learning the structure of Tensorflow codebase.

sanjoy replied:
I'd create a `msvc_compatibility.cc/h` (or `windows_compatibility.cc/h`, if that's better) and put these declarations in there, guarded under `_MSC_VER`.

rongjiecomputer replied:
Done.

sanjoy replied:
I'd avoid putting the `_MSC_VER` stuff here at all -- instead guard the entire contents of `windows_compatibility.cc/h` undef `_MSC_VER`.

Secondly, is `_MSC_VER` the right macro to use for this?  Will it work do the right thing if we build with clang on windows?

rongjiecomputer committed:
Move sincos[f] to windows_compatibility.h

tensorflow/compiler/xla/service/cpu/simple_orc_jit.cc
tensorflow/compiler/xla/service/cpu/windows_compatibility.h

sanjoy commented on tensorflow/compiler/xla/service/cpu/windows_compatibility.h:
```diff
@@ -0,0 +1,35 @@
+
+extern "C" {
+
+void sincos(double x, double *sinv, double *cosv) {
```
These should go into a .cc file with just the declarations in the header.  That way multiple `#include`s of `windows_compatibility.h` will not cause linker errors.

rongjiecomputer replied:
> I'd avoid putting the _MSC_VER stuff here at all -- instead guard the entire contents of windows_compatibility.cc/h undef _MSC_VER.

Done

> Secondly, is _MSC_VER the right macro to use for this? Will it work do the right thing if we build with clang on windows?

Yes, it will work. `clang-cl.exe` tries very hard to pretend it is MSVC. I do like to build Tensorflow with Clang on Windows as well to see if it yields better performance.

sanjoy commented on tensorflow/compiler/xla/service/cpu/BUILD:
```diff
@@ -149,7 +150,7 @@ cc_library(
 cc_library(
     name = "simple_orc_jit",
-    srcs = ["simple_orc_jit.cc"],
+    srcs = ["simple_orc_jit.cc"] + if_windows(["windows_compatibility.h"]),
```
Once you guard the whole file under `_MSC_VER` I think you'll also have to remove the `if_windows` bit and unconditionally list `windows_compatibility.h` in `srcs`.

rongjiecomputer replied:
Done

rongjiecomputer committed:
Move implementation to windows_compatibility.cc

rongjiecomputer committed:
Sort srcs

sanjoy commented on tensorflow/compiler/xla/service/cpu/windows_compatibility.h:
```diff
@@ -0,0 +1,30 @@
+#ifdef _MSC_VER
+
+extern "C" {
+
```
Add a comment here about why we need this.

rongjiecomputer replied:
Done

sanjoy reviewed (approved):
LGTM with that one comment addressed.

rongjiecomputer committed:
Add comment

drpngx commented:
Jenkins, test this please.
